src/caches/boundarympscache.jl

```diff
@@ -0,0 +1,506 @@
+)
+  return switch_messages(bmpsc, partitionpair)
+end
+
+default_niters(alg::Algorithm"orthogonal") = 25
+default_niters(alg::Algorithm"biorthogonal") = 3
+default_tolerance(alg::Algorithm"orthogonal") = 1e-10
+default_tolerance(alg::Algorithm"biorthogonal") = nothing
+
+function default_costfunction(
+  alg::Algorithm"orthogonal",
+  bmpsc::BoundaryMPSCache,
+  pe::PartitionEdge,
+  me::Vector{ITensor},
+)
+  return region_scalar(bp_cache(bmpsc), src(pe)) / norm(only(me))
+end
+
+function default_costfunction(
+  alg::Algorithm"biorthogonal",
+  bmpsc::BoundaryMPSCache,
+  pe::PartitionEdge,
+  me::Vector{ITensor},
+)
+  return region_scalar(bp_cache(bmpsc), src(pe)) /
+         dot(only(me), only(message(bmpsc, reverse(pe))))
+end
+
+#Update all the message tensors on an interpartition via a specified fitting procedure 
+#TODO: Make two-site possible
+function update(
+  alg::Algorithm,
+  bmpsc::BoundaryMPSCache,
+  partitionpair::Pair;
+  inserter=default_inserter,
+  costfunction=default_costfunction,
+  updater=default_updater,
+  cache_prep_function=default_cache_prep_function,
+  niters::Int64=default_niters(alg),
+  tolerance=default_tolerance(alg),
+  normalize=true,
+)
+  bmpsc = cache_prep_function(alg, bmpsc, partitionpair)
+  pes = planargraph_partitionpair_partitionedges(bmpsc, partitionpair)
+  update_seq = vcat(pes, reverse(pes)[2:length(pes)])
+  prev_v, prev_pe = nothing, nothing
+  prev_cf = 0
+  for i in 1:niters
+    cf = 0
+    for update_pe in update_seq
+      cur_v = parent(src(update_pe))
+      bmpsc = updater(alg, bmpsc, prev_pe, update_pe, prev_v, cur_v)
+      me = updated_message(bmpsc, update_pe; message_update_function_kwargs=(; normalize))
+      cf += costfunction(alg, bmpsc, update_pe, me)
+      bmpsc = inserter(alg, bmpsc, update_pe, me)
+      prev_v, prev_pe = cur_v, update_pe
+    end
+    epsilon = abs(cf - prev_cf) / length(update_seq)
+    if !isnothing(tolerance) && epsilon < tolerance
+      return cache_prep_function(alg, bmpsc, partitionpair)
+    else
+      prev_cf = cf
+    end
+  end
+  return cache_prep_function(alg, bmpsc, partitionpair)
+end
+
+#Assume all vertices live in the same partition for now
+function ITensorNetworks.environment(bmpsc::BoundaryMPSCache, verts::Vector; kwargs...)
+  vs = parent.((partitionvertices(bp_cache(bmpsc), verts)))
+  pv = only(planargraph_partitions(bmpsc, vs))
+  bmpsc = partition_update(bmpsc, pv)
+  return environment(bp_cache(bmpsc), verts; kwargs...)
+end
+
+function region_scalar(bmpsc::BoundaryMPSCache, partition)
+  partition_vs = planargraph_vertices(bmpsc, partition)
+  bmpsc = partition_update(bmpsc, first(partition_vs), last(partition_vs))
+  return region_scalar(bp_cache(bmpsc), PartitionVertex(last(partition_vs)))
+end
+
+function region_scalar(bmpsc::BoundaryMPSCache, partitionpair::Pair)
+  pes = planargraph_partitionpair_partitionedges(bmpsc, partitionpair)
+  out = ITensor(1.0)
```

Comment:
```suggestion
  out = ITensor(one(Bool))
```
to be more agnostic about the element type, i.e. `1.0` is implicitly `Float64` which would promote everything to double precision, which isn't good to do on GPUs.